ConcurrentAPI.adoc

```diff
@@ -0,0 +1,220 @@
+== Distributed Concurrent API
+
+=== Motivation
+
+Coming from an initial issue where we need to provide an implementation for vert.x cluster manager for the shared data feature, provide concurrent and distributed programming capabilities.
+
+* Distributed locks -> is it a good idea ?
+* Distributed semaphores : first use case, vert-x cluster management
+* Distributed Countdown latch : it seems the team has already spoken about having it
+* Distributed Atomic Variables : we already have them, would be worth to offer an specific AtomicLong (for example) ?
+
+=== Scope
+
+All these interfaces should be placed in commons, in order to be used either in Embedded and Server mode.
+
+== Infinispan Distributed Locks
+
+I'm wondering if we need to expose Locks on a API. Not sure, because we can already Lock a key on the Cache API and I believe this could be confusing.
+Not sure we need to add anything else right now.
+
+```java
+
+public interface InfinispanLock extends Lock {
+
+}
+
+```
+
+If we decide to do it, the actual implementation could be a wrapper on JGroups Locks.
+
+== Infinipan Distributed Semaphores
+
+Distributed version of java.util.concurrent.Semaphore.
+
+http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html
+
+To create a Semaphore, use the Infinispan Utility class, passing the semaphoreConfig or using the default values or XML values.
+Configuration can be ovrerriden programatically.
+
+First Semaphore wont't implement fairness, this can be cosidered after.
+
+```java
+
+public class Infinispan {
+...
+
+   InfinspanSemaphore semaphore();
+   InfinspanSemaphore semaphore(SemaphoreConfig semaphoreConfig);
+
+}
+
+public class SemaphoreConfig {
+   private String name;
+   private int permits;
+   private boolean createIfAbsent;
+   private boolean silentFailover;
+}
+
+ * @name A unique for the Semaphore
+ * @permits number of permits assigned
+ * @createIfAbsent if the semaphore does not exist, and the flag is set to true, then create it with the name and permits passed
+ * @silentFailover When a node acquired a permit and goes down, if the flag is true, the error will be ignored, the permit released and other nodes will be able to adquire it.
+ * When set to false, all the threads waiting will be interrupted.
+
+```
+
+```java
+
+public interface InfinispanSemaphore {
+
+   void acquire();
+
+   void acquire(int permits);
+
+   void acquireUninterruptibly();
+
+   void acquireUninterruptibly(int permits);
+
+   int availablePermits();
+
+   int drainPermits();
+
+   void reducePermits(int reduction)
+
+   void release();
+
+   void release(int permits);
```

Comment:
For Vert.x it's important that those methods can be called by any thread, not just the thread which acquired the permits.